### PR TITLE
adding additional features

### PR DESCRIPTION
**Pull request by iruzevic**
- Added develop mode variable
- Added custom Js events for hooking logic in the form.js
- Added custom JS filters for changing the default values in the form.js
- Adding admin option to prevent scroll to error and scroll to global error.
- Linting
- Commenting
- Added new file with the variables, filters and events code examples
- Added option to reset form after success submit and hook to override this.
- Added form type selector
- Added default values for creating block with inner block so it is not empty like radios, checkboxes and select.
- Little change to textarea style

https://user-images.githubusercontent.com/23283324/139453520-9ec2d692-167e-4d84-9151-73f1d9c6eaf8.mov

### CODE REVIEW

File: src/Blocks/components/form/assets/form.js [hunk 1]
```diff
@@ -275,9 +359,17 @@ export class Form {
...
+	// Scroll to specific element.
 	scrollToElement = (element) => {
 		if (element !== null) {
-			element.scrollIntoView(true);
+			element.scrollIntoView({block: 'start', behavior: 'smooth'});
 		}
 	}
+
+	// Dispatch custom event.
+	addEvent(element, name) {
```
Comment (volfkarlo) [inline src/Blocks/components/form/assets/form.js:370]: it's a bit weird naming.

Would `dispatchFormEvent` be more appropriate?

**Review by volfkarlo — CHANGES_REQUESTED** [review]
Micro changes and suggestions 😄

File: src/Blocks/components/form/assets/form.js [hunk 2]
```diff
@@ -111,39 +166,44 @@ export class Form {
...
+			// Build data object.
 			const data = {
 				name,
 				value,
 				type,
 			};
 
+			// If checkbox/radio on empty change to empty value.
 			if ((type === 'checkbox' || type === 'radio') && !checked) {
 				data.value= '';
```
Comment (goranalkovic-infinum) [inline src/Blocks/components/form/assets/form.js:186]:
```suggestion
				data.value = '';
```

File: src/Blocks/custom/form-selector/form-selector-editor.scss [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// stylelint-disable declaration-no-important
+
+.es-block-form-selector {
+	&__title,
+	&__title-after {
+		text-align: center;
+		font-style: italic;
+		font-size: 14px;
```
Comment (goranalkovic-infinum) [inline src/Blocks/custom/form-selector/form-selector-editor.scss:8]: Doesn't need to be base-font-size calc'd, but would be nice if it was in `rem`s 😄

File: src/Hooks/Form.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import domReady from '@wordpress/dom-ready';
+
+domReady(() => {
+	const form = document.querySelector('.js-es-block-form');
+
+	if (form) {
+		form.addEventListener('esFormsAfterFormSubmit', (event) => {
+			// Do you logic here.
+		});
+	}
```
Comment (goranalkovic-infinum) [inline src/Hooks/Form.md:31]:
```suggestion
	form?.addEventListener('esFormsAfterFormSubmit', (event) => {
		// Do you logic here.
	});
```

File: src/Hooks/Form.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+domReady(() => {
+	const forms = document.querySelectorAll('.js-es-block-form');
+
+	if (forms.length) {
+		[...forms].forEach((form) => {
+			form.addEventListener('esFormsAfterFormSubmit', (event) => {
+				// Do you logic here.
+			});
+		});
+	}
```
Comment (goranalkovic-infinum) [inline src/Hooks/Form.md:48]:
```suggestion
	const forms = document.querySelectorAll('.js-es-block-form');

	if (!forms.length) {
		return;
	}

	[...forms].forEach((form) => {
		form.addEventListener('esFormsAfterFormSubmit', (event) => {
			// Do you logic here.
		});
	});
```

**Review by goranalkovic-infinum — CHANGES_REQUESTED** [review]
Left some minor comments, it's a _Change requested_ because of the console log.

File: src/Blocks/custom/checkboxes/components/checkboxes-editor.js [hunk 6]
```diff
@@ -21,6 +24,7 @@ export const CheckboxesEditor = ({ attributes, setAttributes, clientId }) => {
 				checkboxesContent: <InnerBlocks
 														allowedBlocks={(typeof checkboxesAllowedBlocks === 'undefined') || checkboxesAllowedBlocks}
 														templateLock={false}
+														template={template}
```
Comment (mbmjertan) [inline src/Blocks/custom/checkboxes/components/checkboxes-editor.js:27]: Perhaps this is just GitHub, but check indentation here

File: src/Blocks/custom/form-selector/components/form-selector-editor.js [hunk 7]
```diff
@@ -1,16 +1,109 @@
...
-		<InnerBlocks
-			allowedBlocks={(typeof formSelectorAllowedBlocks === 'undefined') || formSelectorAllowedBlocks}
-			templateLock={false}
-		/>
+		<>
+			{!hasInnerBlocks &&
+				<>
+					<div className={blockClass}>
+						<div className={`${blockClass}__title`}>
+							{__('Select our created forms', 'eightshift-forms')}
```
Comment (mbmjertan) [inline src/Blocks/custom/form-selector/components/form-selector-editor.js:72]: Perhaps something like this?
```suggestion
							{__('Select one of our templates', 'eightshift-forms')}
```

File: src/Blocks/custom/form-selector/components/form-selector-editor.js [hunk 8]
```diff
@@ -1,16 +1,109 @@
...
+										<IconLabel
+											icon={icons.options}
+											label={sprintf(__('%s form', 'eightshift-forms'), label)}
+										/>
+									</Button>
+								)
+							})}
+						</div>
+						<div className={`${blockClass}__title-after`}>
+							{__('or start building your own', 'eightshift-forms')}
```
Comment (mbmjertan) [inline src/Blocks/custom/form-selector/components/form-selector-editor.js:97]:
```suggestion
							{__('or start building your own form', 'eightshift-forms')}
```

File: src/Blocks/components/form/assets/form.js [hunk 9]
```diff
@@ -243,18 +324,21 @@ export class Form {
...
+	// Submit GTM event.
 	gtmSubmit(element) {
 		const eventName = element.getAttribute('data-tracking-event-name');
 
 		if (eventName) {
 			const gtmData = this.getGtmData(element, eventName);
 
 			if (window?.dataLayer && gtmData?.event) {
+				this.addEvent(element, 'BeforeGtmDataPush');
 				window?.dataLayer.push(gtmData);
```
Comment (dingo-d) [inline src/Blocks/components/form/assets/form.js:336]: Is the `?.` here necessary if you've checked this in the conditional?

File: src/Blocks/custom/radios/components/radios-editor.js [hunk 10]
```diff
@@ -20,6 +24,7 @@ export const RadiosEditor = ({ attributes, setAttributes, clientId }) => {
 				radiosContent: <InnerBlocks
 												allowedBlocks={(typeof radiosAllowedBlocks === 'undefined') || radiosAllowedBlocks}
 												templateLock={false}
+												template={template}
 											/>
```
Comment (mbmjertan) [inline src/Blocks/custom/radios/components/radios-editor.js:28]: Check indentation here

File: src/Blocks/components/form/assets/form.js (hunk 1 above)
Comment (dingo-d) [reply src/Blocks/components/form/assets/form.js:370]: I agree, dispatch sounds better, given that it dispatches the event.

File: src/Blocks/components/form/manifest.json [hunk 11]
```diff
@@ -44,6 +44,18 @@
...
+			"default": true
+		},
+		"formDisableScrollToFieldOnError": {
+			"type": "bool",
+			"default": false
+		},
+		"formDisableScrollToGlobalMessageOnSuccess": {
+			"type": "bool",
+			"default": false
+		},
```
Comment (dingo-d) [inline src/Blocks/components/form/manifest.json:58]: The type should be `boolean`, not `bool` IIRC.

File: src/Blocks/custom/checkboxes/components/checkboxes-editor.js (hunk 6 above)
Comment (dingo-d) [inline src/Blocks/custom/checkboxes/components/checkboxes-editor.js:27]: Spacing here seems waaaay odd

File: src/Blocks/custom/form-selector/components/form-selector-editor.js (hunk 7 above)
Comment (dingo-d) [inline src/Blocks/custom/form-selector/components/form-selector-editor.js:72]:
```suggestion
							{__('Select the pre-built forms', 'eightshift-forms')}
```

File: src/Blocks/custom/greenhouse/greenhouse.php [hunk 12]
```diff
@@ -41,6 +43,15 @@
 		[
 			'formPostId' => $greenhouseFormPostId,
 			'formType' => SettingsGreenhouse::SETTINGS_TYPE_KEY,
+			'formResetOnSuccess' => (bool) !Variables::isDevelopMode(),
```
Comment (dingo-d) [inline src/Blocks/custom/greenhouse/greenhouse.php:46]: No need to cast this to boolean, as the return value is set to boolean and the strict mode will ensure no type juggling happens 🙂

File: src/Blocks/custom/mailchimp/mailchimp.php [hunk 13]
```diff
@@ -41,6 +43,15 @@
 		[
 			'formPostId' => $mailchimpFormPostId,
 			'formType' => SettingsMailchimp::SETTINGS_TYPE_KEY,
+			'formResetOnSuccess' => (bool) !Variables::isDevelopMode(),
```
Comment (dingo-d) [inline src/Blocks/custom/mailchimp/mailchimp.php:46]: Here as well

File: src/Blocks/custom/radios/components/radios-editor.js [hunk 14]
```diff
@@ -20,6 +24,7 @@ export const RadiosEditor = ({ attributes, setAttributes, clientId }) => {
 				radiosContent: <InnerBlocks
 												allowedBlocks={(typeof radiosAllowedBlocks === 'undefined') || radiosAllowedBlocks}
 												templateLock={false}
+												template={template}
```
Comment (dingo-d) [inline src/Blocks/custom/radios/components/radios-editor.js:27]: Odd spacing

File: src/Blocks/custom/select/components/select-editor.js [hunk 15]
```diff
@@ -21,6 +24,7 @@ export const SelectEditor = ({ attributes, setAttributes, clientId }) => {
 				selectOptions: <InnerBlocks
 												allowedBlocks={(typeof selectAllowedBlocks === 'undefined') || selectAllowedBlocks}
 												templateLock={false}
+												template={template}
```
Comment (dingo-d) [inline src/Blocks/custom/select/components/select-editor.js:27]: Odd spacing

File: src/Settings/Settings/SettingsGeneral.php [hunk 16]
```diff
@@ -143,22 +153,41 @@ public function getSettingsGlobalData(): array
...
+						'checkboxLabel' => __('Disable default scripts', 'eightshift-forms'),
 						'checkboxIsChecked' => $this->getOptionValue(self::SETTINGS_GENERAL_DISABLE_DEFAULT_SCRIPTS_KEY) === 'true',
 						'checkboxValue' => 'true',
 					]
 				]
 			],
+			[
+				'component' => 'checkboxes',
+				'checkboxesFieldLabel' => \__('Form behavior', 'eightshift-forms'),
+				'checkboxesFieldHelp' => \__('Set all form behavior options in one place', 'eightshift-forms'),
```
Comment (dingo-d) [inline src/Settings/Settings/SettingsGeneral.php:173]: This looks odd 🤔 In other places you didn't prefix the `__()` function, so you can omit it here as well.

**Review by dingo-d — CHANGES_REQUESTED** [review]
Left a few comments, overall great work! 👍🏼

File: src/Settings/Settings/SettingsGeneral.php (hunk 16 above)
Comment (iruzevic) [reply src/Settings/Settings/SettingsGeneral.php]: i have remove all \__ and replaced it with __ on the whole project

**Review by goranalkovic-infinum — APPROVED** [review]

**Review by mbmjertan — APPROVED** [review]

**Review by dingo-d — APPROVED** [review]

**Pull request merged**